### PR TITLE
refactor(plugins): Start initial work to move to lazy

### PR DESCRIPTION
**Pull request by Uzaaft**
Currently WIP. Will need to test it on various languages before I move this away from draft

### CODE REVIEW

**Comment by abzcoding** [comment]
thanks mate, just make sure not to replace the plugin names in the process 😅

**Comment by Uzaaft** [comment]
> thanks mate, just make sure not to replace the plugin names in the process 😅

Ehhhhh too late haha. But I fixed that now

**Comment by luxus** [comment]
great @Uzaaft thank you so much

**Comment by Uzaaft** [comment]
> great @Uzaaft thank you so much

Had to step up and contribute after having you guys doing so much 😅

**Comment by Uzaaft** [comment]
After the initial migration, we should leverage the folder based system that Lazy.nvim provides to further simplify the dots, if you'd like @abzcoding

**Comment by abzcoding** [comment]
> After the initial migration, we should leverage the folder based system that Lazy.nvim provides to further simplify the dots, if you'd like @abzcoding

hmm, sounds like a good idea,
what would you suggest?

**Comment by Uzaaft** [comment]
> > After the initial migration, we should leverage the folder based system that Lazy.nvim provides to further simplify the dots, if you'd like abzcoding
> 
> hmm, sounds like a good idea, what would you suggest?

I think the best would be to move all plugin specific things in lua/user/* to lua/plugins/*, if I understand correctly. I had a MVP as my nvim conf, so I'll take a look as soon as I'm finished with the initial migration.

**Comment by Uzaaft** [comment]
Things should be stable enough for you guys to test @abzcoding  @luxus 

Had some initial trouble with go(reminds me why I stopped using it xD), so some extra scrutiny of go things would be preferred.

**Comment by Uzaaft** [comment]
Do ping me whenever you have tested it. Somehow, lvim got really slow for me. About 200-220 ms startup times

**Comment by Uzaaft** [comment]
I tested quickly with docker + lvim, and I got a few bugs. Attempted to fix those, but that went nowhere. Just a heads up

**Comment by luxus** [comment]
i tried it and got a lot errors :D

**Comment by Uzaaft** [comment]
> i tried it and got a lot errors :D

Got anything more specific?

**Comment by luxus** [comment]
let me update

**Comment by Uzaaft** [comment]
Sure thing. I found out that I misunderstood the logic of conditionals(we went from disabled to enabled conditions), so I had to fix a few of them. in the most recent commit

**Comment by luxus** [comment]
did i do something wrong? i pulled the lazy pr into the lunarvim and merged your config into my config. but it looks like standard lunarvim

**Comment by Uzaaft** [comment]
Do you have the command `:Lazy` available?

**Comment by luxus** [comment]
okay after restarting and waiting for at least 30 seconds but now it looks better

File: lua/user/plugins.lua
Comment (luxus) [inline lua/user/plugins.lua:162]: you have to use url =

Comment (Uzaaft) [reply lua/user/plugins.lua:162]: Oh good catch.

Comment (Uzaaft) [reply lua/user/plugins.lua:162]: That broke everything for me 😅

**Comment by Uzaaft** [comment]
> 

Is it that slow? :O

File: lua/user/plugins.lua
Comment (luxus) [reply lua/user/plugins.lua:162]: mhh looks like it didn't work.
in a other config       url = "https://git.sr.ht/~whynothugo/lsp_lines.nvim",
works

Comment (Uzaaft) [reply lua/user/plugins.lua:162]: Still broken for me.. I'll have to take a deeper look

**Comment by luxus** [comment]
works now after the 2. start was so slow.
telescope looks strange.. or did the look changed?  
Screenshot 2022-12-28 at 22 38 17

**Comment by Uzaaft** [comment]
> works now after the 2. start was so slow.
> telescope looks strange.. or did the look changed?
> Screenshot 2022-12-28 at 22 38 17

Yeah I agree....Are you using noice and nut?

**Comment by luxus** [comment]
noice

**Comment by Uzaaft** [comment]
> works now after the 2. start was so slow.
> telescope looks strange.. or did the look changed?
> Screenshot 2022-12-28 at 22 38 17

I didn't quite understand how you made it work. Still broken for me with url = ....

**Comment by luxus** [comment]
i simply disabled lsp-lines

**Comment by Uzaaft** [comment]
First priority is to fix the look of noice and nui. I miss the sexy ass looking config.

**Comment by Uzaaft** [comment]
So I went over my work, and reverted most of my attempts to use lazy.nvim's features.

**Comment by abzcoding** [comment]
@Uzaaft thank you very much for your work on this
@luxus thanks for testing it

I'm a bit busy right now, but I'll try it out in a couple of days...

**Comment by Uzaaft** [comment]
No worries

**Comment by LostNeophyte** [comment]
the slowness is because the cache is enabled after reading user's config, I [know how to "fix" it](https://github.com/LunarVim/LunarVim/compare/refactor/lazy...LostNeophyte:LunarVim:lazy-freeze), it's a little bit hacky so I'm waiting for kylo's (or abzcoding's, but I know you're busy) approval before merging it, [here's](https://gist.github.com/LostNeophyte/c3ca8f0325034353d0cd4d3e4e327e45) a script to install the version with cache to /tmp, something may be wrong with that version, the tests are failing

for abzcoding and lexus: you can use lvim appimage form [here](https://github.com/LunarVim/LunarVim-mono/releases/tag/nightly) for testing btw (it doesn't include the cache patch)

**Comment by Uzaaft** [comment]
@luxus Noice should be...(pardon the pun), noice again. Could you tri it again?

**Comment by luxus** [comment]
Yea I will test it around 90 mins, traveling around at the moment

**Comment by luxus** [comment]
lgtm

**Comment by Uzaaft** [comment]
Ai Ai captain 🫡

**Comment by Uzaaft** [comment]
I found it annoying when neotest always loaded all neotest plugins for all languages, so I load them just for a specific language. 
Shaved away some time.

File: lua/user/plugins.lua
Comment (Uzaaft) [inline lua/user/plugins.lua:292]: @abzcoding Ping me If you don't like these, and I'll remove these

Comment (luxus) [reply lua/user/plugins.lua:292]: could it not be loaded on command? with? or did I misunderstood this from the lazy.nvim site

Comment (Uzaaft) [reply lua/user/plugins.lua:292]: Well, I think you're onto something. But does neotest-go and neatest-python have commands? Or are they simply "plugins" that plugs into neotest?

Comment (luxus) [reply lua/user/plugins.lua:292]: Good question! I would use neotest as a command, but this would load all the neotest-plugins.
The BufEnter may be the best solution, but it means neotest will load on .rs, even if you did not intend to run neotest.
Maybe @folke have an idea?

**Comment by luxus** [comment]
> Ai Ai captain 🫡 . I'll be adding lazy = true back whenever we can, to make things super fast

its so much better already <3 thanks for the work.
maybe we can add a load timer in the dashboard? 
https://github.com/lvim-tech/lvim/commit/82b050cfcb6b487a62481b0af5ce6ac6292210ea here is a example

**Comment by Uzaaft** [comment]
Will look into it Luxus.

Bug:
Tab seems to be broken at my end.Whenever I end up tabbing I get:
image

**Comment by luxus** [comment]
i found the timer in lazy itself. so its maybe a "future" enhancement :D

**Comment by Uzaaft** [comment]
Is it a variabel exposed by the api @luxus  ? 
If so, if you gave it I'll add it right away

**Comment by luxus** [comment]
Stats API (require("lazy").stats()):
```lua
{
  -- startuptime in milliseconds till UIEnter
  startuptime = 0,
  -- when true, startuptime is the accurate cputime for the Neovim process. (Linux & Macos)
  -- this is more accurate than `nvim --startuptime`, and as such will be slightly higher
  -- when false, startuptime is calculated based on a delta with a timestamp when lazy started.
  startuptime_cputime = false,
  count = 0, -- total number of plugins
  loaded = 0, -- number of loaded plugins
}
```
`require("lazy").stats().startuptime`
`require("lazy").stats().count`
`require("lazy").stats().loaded`

**Comment by Uzaaft** [comment]
I'm able to get any numbers out from that API @luxus...

**Comment by Uzaaft** [comment]
Bug: Using X in visual mode doesn't seem to work anymore. Will need to look into this.

**Comment by Uzaaft** [comment]
> Bug: Using `x` in visual mode doesn't seem to work anymore. Will need to look into this.

This was due to Hop most likely. At list partially. `x` not works in normal mode, not visual.

**Comment by Uzaaft** [comment]
Issue with tabout is now resolved.
Issue with hop is partially resolved. Still need to iron out the visual mode bug

File: lazy-lock.json
Comment (abzcoding) [inline lazy-lock.json:1]: @Uzaaft can we add this file to gitignore?

File: lua/user/autocommands.lua
Comment (abzcoding) [inline lua/user/autocommands.lua:6]: if we can use this, then remove this please:
https://github.com/abzcoding/lvim/blob/420911a77452ceb3d4c2a27726d83192f2cb3f3c/after/plugin/highlights.lua#L1

File: lua/user/plugins.lua
Comment (abzcoding) [inline lua/user/plugins.lua:382]: shouldn't this be `enabled` instead of `enable`?

File: lua/user/plugins.lua
Comment (abzcoding) [inline lua/user/plugins.lua:624]: Shouldn't this be `enabled` instead of `enable` ?

File: lua/user/plugins.lua
Comment (abzcoding) [inline lua/user/plugins.lua:636]: also this

**Comment by Uzaaft** [comment]
@abzcoding  Thanks for catching those. I blame it on the caffeine withdrawals. Patching those as we speak

File: lua/user/plugins.lua
Comment (Uzaaft) [reply lua/user/plugins.lua]: Fixed in f090f8a

File: lua/user/plugins.lua
Comment (Uzaaft) [reply lua/user/plugins.lua]: Fixed in f090f8a

File: lua/user/plugins.lua
Comment (Uzaaft) [reply lua/user/plugins.lua]: Fixed in f090f8a

File: lua/user/autocommands.lua
Comment (Uzaaft) [reply lua/user/autocommands.lua:6]: Fixed in eafcaeb

File: lazy-lock.json
Comment (Uzaaft) [reply lazy-lock.json:1]: We can. Not sure if we should(no idea regarding the pros/cons of adding it to gitignore)

Comment (abzcoding) [reply lazy-lock.json:1]: hmm, I'm always on the latest version of plugins ( to test them out)
so this file would be changed regularly, 

not sure tbh

Comment (Uzaaft) [reply lazy-lock.json:1]: I’ll add it to the gitignore.

Comment (Uzaaft) [reply lazy-lock.json]: Fixed in 97f2d8d.

File: lua/user/plugins.lua
Comment (abzcoding) [reply lua/user/plugins.lua:293]: it should be `rouge8/neotest-rust`

Comment (Uzaaft) [reply lua/user/plugins.lua]: Fixed in d218f26

**Comment by abzcoding** [comment]
One last thing, tree-sitter installation pops up every time for me. Does the same happen to you guys as well?

**Comment by Uzaaft** [comment]
> One last thing, tree-sitter installation pops up every time for me. Does the same happen to you guys as well?

No. But I saw something regarding that somewhere. Let me look around

**Comment by Uzaaft** [comment]
No @abzcoding . Can't recreate locally. 
Do any of your tree sitter stuff fail?

**Comment by luxus** [comment]
i updated it still no errors :D

**Comment by abzcoding** [comment]
Let's wait for https://github.com/LunarVim/LunarVim/pull/3647 to get merged
and we'll merge this afterward,

@Uzaaft, @luxus, thanks for your efforts on this

**Review by abzcoding — APPROVED** [review]

**Comment by LostNeophyte** [comment]
> One last thing, tree-sitter installation pops up every time for me. Does the same happen to you guys as well?

Someone reported that on my PR, I'll check it on Thursday, but I didn't see that myself, maybe I reinstalled lvim after every use

**Comment by abzcoding** [comment]
> > One last thing, tree-sitter installation pops up every time for me. Does the same happen to you guys as well?
> 
> Someone reported that on my PR, I'll check it on Thursday, but I didn't see that myself, maybe I reinstalled lvim after every use

much appreciated 👍

**Comment by Uzaaft** [comment]
> > One last thing, tree-sitter installation pops up every time for me. Does the same happen to you guys as well?
> 
> Someone reported that on my PR, I'll check it on Thursday, but I didn't see that myself, maybe I reinstalled lvim after every use

treesitter auto-install seems broken with lazy. I had issue on my non-lvim config, but I switched to ensure_installed instead, and it worked fine there. Perhaps that's what the guy in your PR also had happening to him?

**Comment by Uzaaft** [comment]
@abzcoding  @luxus  @LostNeophyte  What kind of startup speed do you guys get? Mine is at 130ish ms.

**Comment by Uzaaft** [comment]
> @abzcoding @luxus @LostNeophyte What kind of startup speed do you guys get? Mine is at 130ish ms. While a similar, non-lvim config hovers at around 30ms

**Comment by LostNeophyte** [comment]
> @abzcoding  @luxus  @LostNeophyte  What kind of startup speed do you guys get? Mine is at 130ish ms. 

90ms(empty config) I have a branch that is 1.2x faster, but I'll leave that for a second pr

**Comment by LostNeophyte** [comment]
> > @abzcoding  @luxus  @LostNeophyte  What kind of startup speed do you guys get? Mine is at 130ish ms. 
> 
> 90ms(empty config) I have a branch that is 1.2x faster, but I'll leave that for a second pr

I measured headless quit, lazy reports anywhere from 90 to 140, packer was slower

**Comment by luxus** [comment]
145ms on macos with 105 plugins, I'm sure I could make it quicker... bit 145ms is already so much better

**Comment by Uzaaft** [comment]
Ah alright. I'll take a look at core lunarvim, something doesn't seem quite right with that loading time.

**Comment by luxus** [comment]
thats true, astrovim takes 42ms :D

**Comment by LostNeophyte** [comment]
> Ah alright. I'll take a look at core lunarvim, something doesn't seem quite right with that loading time.

What's the time for you with master lunarvim and clean config?

**Comment by LostNeophyte** [comment]
I removed site dir from rtp, the treesitter issue should be fixed

**Comment by luxus** [comment]
nice performance updates.. 83 is pretty good. i have 50-55ms with 90 plugins in astrovim. 
lunarvim is still way slower.. hopefully it will catch up

**Comment by LostNeophyte** [comment]
> nice performance updates.. 83 is pretty good. i have 50-55ms with 90 plugins in astrovim. lunarvim is still way slower.. hopefully it will catch up

Can you send a link to your config? I need to see why the difference is this big, I won't rest until it's smaller xD

**Comment by luxus** [comment]
sure it's here.
https://github.com/luxus/AstroNvim_user the install instructions are up to date and it can run next to lunarvim

**Comment by LostNeophyte** [comment]
@luxus 
screenshot from dev channel on discord
image

**Comment by luxus** [comment]
great that astrovim was a help,
thats my config so far (yes I have a light theme on daytime)
Screenshot 2023-01-07 at 14 55 37

**Comment by luxus** [comment]
i did some tests with @LostNeophyte our bloated user conf add 85ms :D but it is still at 120ms

**Comment by Uzaaft** [comment]
@luxus  Why the move to AstroVim, if I may ask?

**Comment by LostNeophyte** [comment]
> @luxus Why the move to AstroVim, if I may ask?

2 reasons are in the #lunarvim channel on discord, one of the newest messages, there may be more

**Comment by luxus** [comment]
i have a personal problem how donations at lunarvim is been handled, and I guess it will be never changed.
i think the main contributors should get the fame and money.

astrovim is very stable.. the documentation is great. @mehalter is a great dev (dunno about the others)
astrovim will have less abstraction with v3 (lazy.nvim) and is very easy to configure

I re-adding all the plugins at the moment I liked about abzconfig config and it will be much faster. 
i want to add some new stuff like hydra for some stuff like git, diagnostics. and that's out of scope to add into @abzcoding config

if you have more questions you can add me on discord.. i was recently active in the #lunarvim channel

**Comment by Uzaaft** [comment]
Yeah I also have a problem with those. That's why I prefer to just donate to people directly. Which is why I asked multiple time if @abzcoding will enable donations. xD 

I'll have to take a look at astrovim. Perhaps that's something I might convert to. We'll see, for now I'm sticking to lvim for debugging and my own lazyvim config for other things.

**Comment by luxus** [comment]
yeah i asked abzconfig about donations, too .. :D 

with @LostNeophyte improvements lvim is very usable; I'm only sad that I couldn't recommend lunarvim anymore. that's why I try to get astrovim bloated to be happy again :D

**Comment by Uzaaft** [comment]
Yeah, LostNeoPhyte has done some amazing improvements to lvim. I just like things when they are very fast( **Blazingly fast** ), and 30ms vs 120ms is a no-brainer currently. xD

**Comment by luxus** [comment]
keep in mind that abzcoding could drag even astrovim down :D

**Comment by Uzaaft** [comment]
I'll report back when my config is 1:1 with abz's xD

**Comment by luxus** [comment]
you can check my config, i added already a lot bloat :D

**Comment by Uzaaft** [comment]
@luxus  They have removed the donation links in the Lunarvim repo now

**Comment by abzcoding** [comment]
@Uzaaft, tyvm for your work on this ❤️ 
@luxus, thanks for testing this out 👍 
@LostNeophyte keep on making lunarvim better 💪

**Pull request merged**
